src/common/runnable/run-many-for.ts

```diff
@@ -3,46 +3,81 @@
  * Licensed under MIT License. See LICENSE in root directory for more information.
  */
 import type { DiContainerForInjection, InjectionToken } from "@ogre-tools/injectable";
-import type { Composite } from "../utils/composite/get-composite/get-composite";
-import { getCompositeFor } from "../utils/composite/get-composite/get-composite";
+import type { SingleOrMany } from "../utils";
+import { getOrInsertSet, isDefined } from "../utils";
+import { observable, when } from "mobx";
 import * as uuid from "uuid";
+import assert from "assert";
 
 export interface Runnable<TParameter = void> {
   id: string;
   run: Run<TParameter>;
-  runAfter?: Runnable<TParameter>;
+  runAfter?: SingleOrMany<Runnable<TParameter>>;
 }
 
 type Run<Param> = (parameter: Param) => Promise<void> | void;
 
 export type RunMany = <Param>(injectionToken: InjectionToken<Runnable<Param>, void>) => Run<Param>;
 
-async function runCompositeRunnables<Param>(param: Param, composite: Composite<Runnable<Param>>) {
-  await composite.value.run(param);
-  await Promise.all(composite.children.map(composite => runCompositeRunnables(param, composite)));
-}
+const computedNextEdge = (traversed: string[], graph: Map<string, Set<string>>, currentId: string) => {
+  const currentNode = graph.get(currentId);
+
+  assert(currentNode, `Runnable graph does not contain node with id="${currentId}"`);
+
+  for (const nextId of currentNode.values()) {
+    if (traversed.includes(nextId)) {
+      throw new Error(`Cycle in runnable graph: "${traversed.join(`" -> "`)}" -> "${nextId}"`);
+    }
+
+    computedNextEdge([...traversed, nextId], graph, nextId);
+  }
+};
+
+const verifyRunnablesAreDAG = <Param>(runnables: Runnable<Param>[]) => {
+  const rootId = uuid.v4();
+  const runnableGraph = new Map<string, Set<string>>();
+
+  // Build the Directed graph
+  for (const runnable of runnables) {
+    getOrInsertSet(runnableGraph, runnable.id);
+
+    if (!runnable.runAfter) {
+      getOrInsertSet(runnableGraph, rootId).add(runnable.id);
+    } else if (Array.isArray(runnable.runAfter)) {
+      for (const parentRunnable of runnable.runAfter) {
+        getOrInsertSet(runnableGraph, parentRunnable.id).add(runnable.id);
+      }
+    } else {
+      getOrInsertSet(runnableGraph, runnable.runAfter.id).add(runnable.id);
+    }
+  }
+
+  // Do a DFS to find any cycles
+  computedNextEdge([], runnableGraph, rootId);
+};
+
+const executeRunnableWith = <Param>(param: Param) => {
+  const finishedIds = observable.set<string>();
+
+  return async (runnable: Runnable<Param>): Promise<void> => {
+    const parentRunnables = [runnable.runAfter].flat().filter(isDefined);
+
+    for (const parentRunnable of parentRunnables) {
+      await when(() => finishedIds.has(parentRunnable.id));
```

Comment:
For this algorithm, using `MobX` is a bit too exotic for my preference, and most likely is the wrong tool for the job. Instead of `when`, could there instead be eg. a `Map` with `runnable.id` as key and the related promise as value?

Reply:
Will try other solutions

Reply:
That by itself isn't enough because the order of the runnables are not guaranteed to be in the correct order for that insertion to be guaranteed to happen.

Reply:
Done